### PR TITLE
feat(skore): Add infrastructure to support different plotting backends for Displays

### LINKED ISSUE #1079: feat(skore): Add plotly as plotting backend for `Display.plot`
### Is your feature request related to a problem? Please describe.

*This was also requested in a user interview. A DS consultant needs plots with interaction to answer questions from clients during a presentation.*

`EstimatorReport` has some plotting features, but only in matplotlib for now. Hence, in the skore UI, the display would not be neat.

### Describe the solution you'd like

We should add plotly as a plotting backend. Kind of what Pandas allows: [Pandas Plotting Backend in Python](https://plotly.com/python/pandas-backend/).

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by divakaivan**
Closes #1079

### CODE REVIEW

**Comment by divakaivan** [comment]
`make test` output:  `926 passed, 5 skipped, 909 warnings`

File: skore/src/skore/_config.py [hunk 1]
```diff
@@ -61,6 +63,14 @@ def set_config(
     show_progress : bool, default=None
         If True, show progress bars. Otherwise, do not show them.
 
+    plot_backend : str, default=None
+        The default plotting backend to be used.
+
+        - `"default"`: Default backend for the type of plot
+        - `"matplotlib"`: Use Matplotlib for plotting
+        - `"plotly"`: Use Plotly for plotting
+        - `None`: Plotting backend is unchanged
```
Comment (thomass-dev) [inline skore/src/skore/_config.py]:
```python
# skore/src/skore/_sklearn/_plot/utils.py
class PlotBackendMixin:
    def plot(self, **kwargs):
        plot_backend = get_config()["plot_backend"]
        if plot_backend in ["default", "matplotlib"]:
            return self._plot_matplotlib(**kwargs)
[...]            
```

As i can see in the above snippet, `default` and `matplotlib` are exchangeable.
The `default` cannot be finely defined for each type of display.
So i don't se the point of `default` compared to using `matplotlib`.

IMO, `default` should be removed and replaced by `matplotlib`.

```suggestion
    plot_backend : str | None, default=None
        The plotting backend to be used.

        - `"matplotlib"`: Use Matplotlib for plotting
        - `"plotly"`: Use Plotly for plotting
        - `None`: Plotting backend is unchanged
```

File: skore/src/skore/_config.py [hunk 2]
```diff
@@ -49,6 +50,7 @@ def get_config() -> dict[str, Any]:
 
 def set_config(
     show_progress: bool | None = None,
+    plot_backend: str | None = None,
```
Comment (thomass-dev) [inline skore/src/skore/_config.py]:
```suggestion
    *,
    show_progress: bool | None = None,
    plot_backend: str | None = None,
```

File: skore/src/skore/_config.py [hunk 3]
```diff
@@ -70,18 +80,21 @@ def set_config(
...
     >>> # xdoctest: +SKIP
     >>> from skore import set_config
-    >>> set_config(show_progress=False)
+    >>> set_config(show_progress=False, plot_backend="plotly")
     """
     local_config = _get_threadlocal_config()
 
     if show_progress is not None:
         local_config["show_progress"] = show_progress
+    if plot_backend is not None:
```
Comment (thomass-dev) [inline skore/src/skore/_config.py:94]: Nitpick: as `None` is not a valid value for `plot_backend` in this context, we should use ellipsis as a sentinel.

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py [hunk 4]
```diff
@@ -837,6 +838,13 @@ def plot(
             if despine:
                 _despine_matplotlib_axis(self.ax_)
 
+    @StyleDisplayMixin.style_plot
+    def _plot_plotly(self, **kwargs):
+        raise NotImplementedError(
+            "Plots with plotly is not implemented. "
+            "Please use one of ['default', 'matplotlib']."
+        )
```
Comment (thomass-dev) [inline skore/src/skore/_sklearn/_plot/metrics/roc_curve.py]: Why is this necessary (identical for each metric), when mixin defines this same default behavior?

**Review by thomass-dev — CHANGES_REQUESTED** [review]
Nice! The overall structure LGTM, just few thoughts.

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py (hunk 4 above)
Comment (divakaivan) [reply skore/src/skore/_sklearn/_plot/metrics/roc_curve.py]: Yes. I added the Mixin class after and forgot to remove ~ Will remove

File: skore/src/skore/_config.py (hunk 1 above)
Comment (divakaivan) [reply skore/src/skore/_config.py]: At the moment default and matplotlib do the same thing. I think the idea was that when `plot_backend` is untouched/set to `default`, some displays will use matplotlib and some plotly (whichever we choose to be best for a particular display). Ultimately ~ `default` will allow to use both type of backends for different displays. Would something like this be better, **if** we want this behaviour of using both type of plots:

For example:

```python
### set an attribute in each class for its default plot backend
class ConfusionMatrixDisplay(PlotBackendMixin, Display):
    def __init__(self, ...):
        self._default_plot_backend = self._plot_matplotlib


class PlotBackendMixin:
    if plot_backend == "default":
        return self._default_plot_backend(**kwargs)
    elif plot_backend == "matplotlib":
        return self._plot_matplotlib(**kwargs)
    elif plot_backend == "plotly":
        return self._plot_plotly(**kwargs)
    else: ...
```

File: skore/src/skore/_config.py (hunk 3 above)
Comment (divakaivan) [reply skore/src/skore/_config.py:94]: 1. I can get the mypy to be happy if I set 
```python
def set_config(..., plot_backend: str | Any = ...:
    if plot_backend is not ...:
        local_config["plot_backend"] = plot_backend
```
But this would be using Any

2. `from types import EllipsisType` ([added in Python 3.10](https://stackoverflow.com/questions/66636262/get-python-ellipsis-type-without-having-to-define-it-as-type))
```python
def set_config(..., plot_backend: str | EllipsisType = ...):
    if plot_backend is not ...:
        local_config["plot_backend"] = plot_backend
```

3.  only accepting `str`:
```python
def set_config(..., plot_backend: str = "default"):
    if plot_backend != "default":
        local_config["plot_backend"] = plot_backend
```

File: skore/src/skore/_config.py
Comment (auguste-probabl) [inline skore/src/skore/_config.py:1]: Define a new type `PlotBackend = Literal["default", "matplotlib", "plotly"]`

File: skore/src/skore/_config.py (hunk 1 above)
Comment (auguste-probabl) [reply skore/src/skore/_config.py]: I also think that `default` is superfluous as it feels indistinguishable from `None`

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 5]
```diff
@@ -242,6 +243,29 @@ def _validate_from_predictions_params(
...
+    def plot(self, **kwargs):
+        plot_backend = get_config()["plot_backend"]
+        if plot_backend in ["default", "matplotlib"]:
+            return self._plot_matplotlib(**kwargs)
+        elif plot_backend == "plotly":
+            return self._plot_plotly(**kwargs)
+        else:
+            raise NotImplementedError(
+                f"Plotting backend {plot_backend} not available. "
+                "Available options are ['default', 'matplotlib', 'plotly']."
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/utils.py]: don't hard-code this; define the correct values in the same place as the new `PlotBackend` type

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 6]
```diff
@@ -242,6 +243,29 @@ def _validate_from_predictions_params(
...
+            )
+
+    def _plot_matplotlib(self, **kwargs):
+        raise NotImplementedError("_plot_matplotlib must be implemented by subclass.")
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/utils.py]:
```suggestion
        raise NotImplementedError("Plotting with matplotlib is not supported.")
```

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 7]
```diff
@@ -242,6 +243,29 @@ def _validate_from_predictions_params(
...
+
+    def _plot_plotly(self, **kwargs):
+        raise NotImplementedError(
+            "Plots with plotly is not implemented. "
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/utils.py]:
```suggestion
            "Plotting with plotly is not supported. "
```

File: skore/src/skore/_sklearn/_plot/utils.py (hunk 5 above)
Comment (divakaivan) [reply skore/src/skore/_sklearn/_plot/utils.py]: I created a `PlotBackend` as suggested, and used `PlotBackend.__args__` to list the options. Found it in [this stackoverflow thread](https://stackoverflow.com/questions/48572831/how-to-access-the-type-arguments-of-typing-generic/50101934#50101934)

File: skore/src/skore/_config.py [hunk 8]
```diff
@@ -48,7 +54,9 @@ def get_config() -> dict[str, Any]:
 
 
 def set_config(
+    *,
```
Comment (auguste-probabl) [inline skore/src/skore/_config.py:57]: nice find!

File: skore/src/skore/_config.py [hunk 9]
```diff
@@ -61,6 +69,13 @@ def set_config(
     show_progress : bool, default=None
         If True, show progress bars. Otherwise, do not show them.
 
+    plot_backend : str, default="matplotlib"
```
Comment (auguste-probabl) [inline skore/src/skore/_config.py]:
```suggestion
    plot_backend : {"matplotlib", "plotly"}, default="matplotlib"
```

File: skore/src/skore/_config.py [hunk 10]
```diff
@@ -61,6 +69,13 @@ def set_config(
...
+        The default plotting backend to be used.
```
Comment (auguste-probabl) [inline skore/src/skore/_config.py]:
```suggestion
        The plotting backend to be used.
```

File: skore/src/skore/_config.py (hunk 3 above)
Comment (auguste-probabl) [reply skore/src/skore/_config.py:94]: I don't see what could go wrong here; just in case, if we do something I'd prefer using `skore._sklearn.types._DEFAULT` rather than ellipsis.

File: skore/src/skore/_config.py [hunk 11]
```diff
@@ -94,6 +112,13 @@ def config_context(
     show_progress : bool, default=None
         If True, show progress bars. Otherwise, do not show them.
 
+    plot_backend : str, default="matplotlib"
```
Comment (auguste-probabl) [inline skore/src/skore/_config.py]: See above

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 12]
```diff
@@ -242,6 +249,28 @@ def _validate_from_predictions_params(
...
+        elif plot_backend == "plotly":
+            return self._plot_plotly(**kwargs)
+        else:
+            raise NotImplementedError(
+                f"Plotting backend {plot_backend} not available. "
+                f"Available options are {PlotBackend.__args__}."
+            )
+
+    def _plot_matplotlib(self, **kwargs):
+        raise NotImplementedError("Plotting with matplotlib is not supported.")
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/utils.py]: Since matplotlib is the preferred implementation for now, I think this base case shouldn't exist -- if _plot_matplotlib is not implemented then something has gone terribly wrong.

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 13]
```diff
@@ -242,6 +249,28 @@ def _validate_from_predictions_params(
...
+
+    def _plot_plotly(self, **kwargs):
+        raise NotImplementedError(
+            "Plotting with plotly is not supported. Please use one of ['matplotlib']."
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/utils.py]:
```suggestion
            "Plotting with plotly is not supported for this Display."
```
just to be clear that it might be supported for other displays. Eventually I guess we'd need a catalogue of which Displays support which plotting backends, and this error message would refer to that.

File: skore/src/skore/_config.py [hunk 14]
```diff
@@ -94,6 +112,13 @@ def config_context(
     show_progress : bool, default=None
         If True, show progress bars. Otherwise, do not show them.
 
+    plot_backend : {"matplotlib", "plotly"}, default="matplotlib"
+        The default plotting backend to be used.
```
Comment (auguste-probabl) [inline skore/src/skore/_config.py]: here also
```suggestion
        The plotting backend to be used.
```
I would like something like DocStrands to synchronize the docstrings...

Comment (divakaivan) [reply skore/src/skore/_config.py]: that's nice. recently published - [DocStrands repo](https://github.com/multimeric/DocStrands)

File: skore/src/skore/_config.py [hunk 15]
```diff
@@ -61,6 +69,13 @@ def set_config(
     show_progress : bool, default=None
         If True, show progress bars. Otherwise, do not show them.
 
+    plot_backend : {"matplotlib", "plotly"}, default="matplotlib"
```
Comment (thomass-dev) [inline skore/src/skore/_config.py]:
```suggestion
    plot_backend : {"matplotlib", "plotly"}, default=None
```

I don't think it is the right place to mention that the default backend is `matplotlib`.

File: skore/src/skore/_config.py [hunk 16]
```diff
@@ -94,6 +112,13 @@ def config_context(
...
```
Comment (thomass-dev) [inline skore/src/skore/_config.py]:
```suggestion
    plot_backend : {"matplotlib", "plotly"}, default=None
```

I don't think it is the right place to mention that the default backend is matplotlib.

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 17]
```diff
@@ -242,6 +249,27 @@ def _validate_from_predictions_params(
...
+        else:
+            raise NotImplementedError(
+                f"Plotting backend {plot_backend} not available. "
+                f"Available options are {PlotBackend.__args__}."
+            )
+
+    def _plot_plotly(self, **kwargs):
+        raise NotImplementedError(
+            "Plotting with plotly is not supported for this Display."
+        )
```
Comment (thomass-dev) [inline skore/src/skore/_sklearn/_plot/utils.py:354]: Just my 2cts @auguste-probabl @divakaivan : after re-reading the PR, i find it quite strange to have both the `Display` protocol and the `PlotBackendMixin` mixin, with almost the same purpose.

What about merging the two in a new abstract class `Display`?

Comment (divakaivan) [reply skore/src/skore/_sklearn/_plot/utils.py:354]: I can see them being merged. Auguste mentioned something about a possible restructuring of the Display later on. I think he’s looking at this from more of a bird’s-eye view than I am so curiouis of his PoV ~

Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_plot/utils.py:354]: I'd like to keep the two separate. The two classes don't serve the same purpose: Display is an interface while PlotBackendMixin is an implementation.

**Comment by auguste-probabl** [comment]
@divakaivan Can you clarify the commit message? This PR doesn't actually make it possible to plot via plotly unfortunately.

File: skore/src/skore/_sklearn/_plot/utils.py (hunk 17 above)
Comment (thomass-dev) [reply skore/src/skore/_sklearn/_plot/utils.py:354]: I disagree. 

As long as there is only one `plot` implementation (what i see), the interface `Display` and the implementation `PlotBackendMixin` have the same purpose: defining the behavior of the display object. Having them in one place improves readability. IMO, there is no problem to have function definition in an interface. Something like: 

```python
from abc import ABC

class Display(ABC):  # DisplayMixin
    def plot(): ...
    def plot_plotly(): raise NotImplementedError
    def plot_matplotlib(): raise NotImplementedError    
    
    @abstractmethod
    def set_style(self, **kwargs): ...
    
    @abstractmethod
    def frame(self, **kwargs): ...
```

It's not blocking, just an architectural question. Feel free to resolve the conversation @auguste-probabl.

**Review by thomass-dev — APPROVED** [review]

**Comment by glemaitre** [comment]
Just catching up here.

Regarding the `"default"` case, it is exactly what @divakaivan mentioned: we want a display to define its default. So why do we need `None`: in case of nested configuration where we need to detect if we set `"default"` that semantically means choose the default of the display or `None` that semantically means "do not change anything".

From what I recall, without the two options, you are not going to properly detect thing in a nested case.

To be explicit:

```python
with config_context(plot_backend=None):  
    with config_context(plot_backend="plotly"):
        print(get_config()["plot_backend"])
```

There is no way that I revert it to the `"default"` backend, while what I wanted is:


```python
with config_context(plot_backend="default"):  
    with config_context(plot_backend="plotly"):
        print(get_config()["plot_backend"])
```

**Comment by glemaitre** [comment]
I open https://github.com/probabl-ai/skore/issues/1951 to avoid continuing the discussion in a merged PR ;)
